### PR TITLE
feat: add event tags, properties, and PostHog AI tracing integration

### PR DESCRIPTION
**Pull request by naji247**
## Summary

This PR introduces two major features for MCPCat's event pipeline: **customer-defined event metadata** (tags + properties) and **PostHog AI observability integration**. It also improves how all four telemetry exporters handle customer data and source attribution.

---

### Problem
Customers had no way to attach custom metadata to the events MCPCat automatically captures. The only workaround was piggybacking on person properties via `identify`, which polluted user profiles with per-event data.

### Solution
Two new callbacks on `MCPCatOptions` that are invoked on every auto-captured event (tool calls, tool lists, initialize):

**`eventTags`** — String key-value pairs intended for indexed, queryable metadata:
```typescript
eventTags: async (request, extra) => ({
  trace_id: extra?.requestContext?.traceId,
  env: process.env.NODE_ENV,
  region: "us-east-1",
})
```
- Validated client-side: keys ≤32 chars matching `[a-zA-Z0-9_.:\- ]`, values ≤200 chars, max 50 entries
- Invalid entries silently dropped with warning logged to `~/mcpcat.log`
- Intended for structured metadata you'd filter/group by in dashboards

**`eventProperties`** — Flexible JSON for richer, unstructured metadata:
```typescript
eventProperties: async (request, extra) => ({
  device: "desktop",
  app_version: "2.1.0",
  feature_flags: ["dark_mode", "beta_ui"],
})
```
- No constraints beyond standard JSON types
- Bypasses redaction/sanitization pipeline (customer explicitly provides this data)

### How it works
1. Callbacks receive the same `(request, extra)` arguments as `identify`
2. Resolved at each instrumentation point (tool calls, initialize, tools/list)
3. Carried through the publish pipeline as top-level `tags`/`properties` fields on `Event`
4. Each exporter maps them to platform-appropriate locations (see Exporter Improvements below)
5. Also available on `publishCustomEvent()` via the `CustomEventData` object (direct, no callback)

### Files changed
- `src/types.ts` — Added `eventTags`/`eventProperties` to `MCPCatOptions`, `tags`/`properties` to `Event` and `CustomEventData`
- `src/modules/validation.ts` — **New file.** Client-side tag validation (key format, length, entry count)
- `src/modules/internal.ts` — `resolveEventTags()` and `resolveEventProperties()` helpers
- `src/modules/tracingV2.ts` — Resolve callbacks in high-level McpServer tool call wrapper
- `src/modules/tracing.ts` — Resolve callbacks in low-level Server handlers (4 paths)
- `src/modules/eventQueue.ts` — Carry tags/properties through publish pipeline
- `src/modules/redaction.ts` — Exempt tags/properties from redaction
- `src/index.ts` — Wire options through `track()`, validate tags in `publishCustomEvent()`

---

### Problem
PostHog's team requested deeper integration with their AI observability pipeline. MCPCat was sending generic capture events (`mcp_tool_call`, etc.) that didn't appear in PostHog's trace views, latency charts, or AI-specific dashboards.

### Solution
A new `enableAITracing` flag on `PostHogExporterConfig` that emits `$ai_span` events for tool calls alongside regular capture events:

```typescript
exporters: {
  posthog: {
    type: "posthog",
    apiKey: "phc_...",
    enableAITracing: true,
  }
}
```

### PostHog property mapping

| PostHog Property | MCPCat Source | Description |
|---|---|---|
| `$ai_session_id` | `"mcpcat_" + sessionId` | Groups all tool calls in a session |
| `$ai_trace_id` | `toUUID(sessionId)` | Deterministic UUID, same session scope |
| `$ai_span_id` | `toUUID(eventId)` | Deterministic UUID, unique per tool call |
| `$ai_span_name` | `event.resourceName` | Tool name (e.g., `get_weather`) |
| `$ai_latency` | `event.duration / 1000` | Converted from ms to seconds |
| `$ai_input_state` | `event.parameters` | Redacted tool call parameters |
| `$ai_output_state` | `event.response` | Redacted tool call response |
| `$ai_is_error` / `$ai_error` | `event.isError` / `event.error` | Error tracking |

### Tags/properties on `$ai_span` events
Customer tags and properties are **spread directly** as top-level properties on `$ai_span` events (not namespaced like on regular capture events). This maximizes queryability in PostHog's AI observability views. Customers can override any `$ai_*` default — for example, supplying a custom `$ai_trace_id` via `eventTags` to control trace grouping.

### How it works
- Additive: regular `mcp_tool_call` events continue to be emitted (preserves existing dashboards)
- Tool calls only: `$ai_span` is not emitted for initialize, list, or other event types
- Single network call: the `$ai_span` is appended to the existing batch request
- Deterministic UUIDs: `toUUID()` generates SHA-256-based UUIDs with v5-compatible formatting, ensuring the same session/event always maps to the same ID

### Files changed
- `src/modules/exporters/posthog.ts` — `enableAITracing` config, `toUUID()` helper, `buildAISpanEvent()` method, conditional in `export()`

---

### Source attribution
All exporters now include `source: "mcpcat"` via a shared `MCPCAT_SOURCE` constant for consistent attribution.

### Datadog
- Customer tags namespaced with `mcpcat.` prefix to avoid collisions with reserved Datadog tags
- Key sanitization: lowercase, strip colons/spaces/commas
- Value sanitization: strip commas (Datadog ddtags delimiter)
- `source:mcpcat` added to ddtags

### Sentry
- Customer properties moved from deprecated `extra` field to `contexts.mcpcat`
- New `buildContexts()` helper for DRY context construction across transactions and error events
- `source: "mcpcat"` added as a Sentry tag
- Customer tags mapped directly to Sentry tags (Sentry tags are already the right level for filtering)

### OTLP
- Customer tags as individual span attributes: `mcpcat.tag.{key}`
- Customer properties as JSON string attribute: `mcpcat.properties`
- `source: "mcpcat"` span attribute added

### PostHog (regular capture events)
- Customer tags as `mcpcat_tag_{key}` properties
- Customer properties as `mcpcat_properties` nested object
- `source: "mcpcat"` property added
- Extracted `getDistinctId()`, `getTimestamp()`, `toUUID()` as shared module-level helpers

---

## Code quality
- `MCPCAT_SOURCE` constant in `src/modules/constants.ts` replaces 6 hardcoded `"mcpcat"` strings across 4 exporters
- Shared helpers eliminate `distinct_id`/`timestamp` duplication across 3 PostHog event builders
- `toUUID()` extracted as pure module-level function (no instance state)

---

## Test coverage

**27 new tests** across 3 test files:

- `src/tests/validation.test.ts` — 11 unit tests for tag validation (key format, length, value constraints, entry limit, edge cases)
- `src/tests/event-tags-properties.test.ts` — 16 integration tests (tool calls, initialize, tools/list, callback args, publishCustomEvent, error resilience, conditional assignment)
- `src/tests/posthog-exporter.test.ts` — 10 new tests (AI span emission, UUID determinism, flag-off guard, non-tool-call guard, property mapping, tags/properties spread, override behavior, error 3-event batch, existing tags/properties namespacing)

**424 tests passing, 1 skipped, typecheck clean.**

## Test plan
- [ ] `pnpm test` — all 424 tests pass
- [ ] `pnpm tsc --noEmit` — no type errors
- [ ] Verify PostHog AI tracing: enable `enableAITracing: true`, make tool calls, confirm `$ai_span` events in PostHog AI observability dashboard with correct `$ai_session_id`, `$ai_trace_id`, `$ai_span_id`, `$ai_span_name`
- [ ] Verify event tags: configure `eventTags` callback, confirm tags appear correctly in each exporter (Sentry tags, Datadog `mcpcat.*` ddtags, OTLP `mcpcat.tag.*` attributes, PostHog `mcpcat_tag_*` properties)
- [ ] Verify event properties: configure `eventProperties` callback, confirm properties in Sentry `contexts.mcpcat`, Datadog `mcp.properties`, OTLP `mcpcat.properties`, PostHog `mcpcat_properties`
- [ ] Verify tag validation: submit invalid tags (long keys, bad chars, >50 entries), confirm dropped with warning in `~/mcpcat.log`
- [ ] Verify `publishCustomEvent()` supports `tags`/`properties` directly on the event data object

### CODE REVIEW

File: src/modules/exporters/datadog.ts [hunk 1]
```diff
@@ -141,6 +143,13 @@ export class DatadogExporter implements Exporter {
     if (event.resourceName) tags.push(`resource:${event.resourceName}`);
```
Comment (kashishhora) [inline src/modules/exporters/datadog.ts:144]: I think these tags might collide with internal Datadog tags. Should we consider prefixing them with `mcpcat_` so they don't collide?

File: src/modules/exporters/sentry.ts [hunk 2]
```diff
@@ -379,6 +379,13 @@ export class SentryExporter implements Exporter {
     if (event.clientName) tags.client_name = event.clientName;
     if (event.identifyActorGivenId) tags.actor_id = event.identifyActorGivenId;
 
+    // Add customer-defined tags
+    if (event.tags) {
```
Comment (kashishhora) [inline src/modules/exporters/sentry.ts:385]: Same comment here regarding potential tag collisions. Thoughts on prepending tags keys with `mcpcat_`?

File: src/modules/exporters/datadog.ts [hunk 3]
```diff
@@ -141,6 +143,13 @@ export class DatadogExporter implements Exporter {
...
     if (event.isError) tags.push("error:true");
 
+    // Add customer-defined tags to ddtags
+    if (event.tags) {
+      for (const [key, value] of Object.entries(event.tags)) {
+        tags.push(`${key}:${value}`);
```
Comment (kashishhora) [inline src/modules/exporters/datadog.ts]: Should comma escape these so they don't get split up by Datadog when it parses this.

File: src/modules/redaction.ts [hunk 4]
```diff
@@ -16,6 +16,8 @@ const PROTECTED_FIELDS = new Set([
   "resourceName",
   "eventType",
   "actorId",
+  "tags",
```
Comment (kashishhora) [inline src/modules/redaction.ts:19]: I added these two to `PROTECTED_FIELDS` so that they don't get redacted. Also added a test to make sure it works.

**Review by kashishhora — APPROVED** [review]

**Comment by naji247** [comment]
## PostHog AI Observability Alignment

This PR addresses the PostHog team's request for deeper integration with their AI observability pipeline. Here's how MCPCat events and tags now support PostHog's tracing model:

### How MCPCat maps to PostHog's `$ai_span` schema

| PostHog concept | MCPCat source | Details |
|---|---|---|
| **Session** (`$ai_session_id`) | MCPCat session ID | `mcpcat_` + session KSUID — groups all tool calls under one umbrella |
| **Trace** (`$ai_trace_id`) | MCPCat session ID | Deterministic UUID from session — same scope as session, PostHog uses this for trace grouping |
| **Span** (`$ai_span_id`) | MCPCat event ID | Deterministic UUID from event — each tool call is a unique span |
| **Span name** (`$ai_span_name`) | Tool resource name | Direct mapping — `get_weather`, `search_docs`, etc. |
| **Latency** (`$ai_latency`) | `event.duration` | Converted from ms to seconds |
| **Input/Output** (`$ai_input_state`/`$ai_output_state`) | Tool parameters/response | Already redacted through the MCPCat pipeline |
| **Error** (`$ai_is_error`/`$ai_error`) | MCPCat error tracking | Direct mapping from error detection |

### How `eventTags` and `eventProperties` support PostHog's needs

The new `eventTags` and `eventProperties` callbacks directly address the request for "properties that are always sent with events":

- **`eventTags`** — String key-value pairs resolved per-request. On `$ai_span` events, tags are spread directly as top-level properties (not namespaced) for maximum queryability in PostHog's AI views. Customers can even override `$ai_*` defaults like `$ai_trace_id` or `$ai_session_id` if they need custom grouping.

- **`eventProperties`** — Flexible JSON resolved per-request. Also spread directly on `$ai_span` events. This replaces the workaround of piggybacking on person properties via `identify`.

### Opting in

```typescript
mcpcat.track(server, projectId, {
  exporters: {
    posthog: {
      type: "posthog",
      apiKey: "phc_...",
      enableAITracing: true,  // ← enables $ai_span emission
    }
  },
  eventTags: (request, extra) => ({
    env: "production",
    team: "platform",
  }),
  eventProperties: (request, extra) => ({
    deployment_id: "abc-123",
    feature_flags: ["new_ui"],
  }),
});
```

With `enableAITracing: true`, every `mcp:tools/call` event emits both the regular `mcp_tool_call` capture event (preserving existing dashboards) and an `$ai_span` event that appears natively in PostHog's AI observability trace views.

**Comment by naji247** [comment]
## Custom Trace, Span, and Session IDs via `eventTags`

When `enableAITracing` is enabled on the PostHog exporter, MCPCat generates deterministic `$ai_trace_id`, `$ai_span_id`, and `$ai_session_id` values from its internal session and event IDs. However, customers can override any of these by returning the corresponding keys from their `eventTags` callback:

```typescript
mcpcat.track(server, projectId, {
  exporters: {
    posthog: {
      type: "posthog",
      apiKey: "phc_...",
      enableAITracing: true,
    }
  },
  eventTags: (request, extra) => ({
    // Override PostHog AI tracing IDs with your own values
    $ai_session_id: myApp.getSessionId(),
    $ai_trace_id: myApp.getTraceId(),
    $ai_span_id: myApp.getSpanId(),

    // You can also override the span name
    $ai_span_name: `${request.params.name}:${myApp.getRegion()}`,
  }),
});
```

### How it works

MCPCat builds the `$ai_span` event properties with its own defaults first, then spreads customer tags on top. Any `$ai_*` key returned from `eventTags` replaces the MCPCat-generated value. This means:

- **`$ai_session_id`** — Default: `mcpcat_<sessionId>`. Override to group tool calls into your own session boundaries.
- **`$ai_trace_id`** — Default: deterministic UUID from MCPCat session ID. Override to correlate MCPCat spans with traces from your own instrumentation (e.g., OpenTelemetry trace IDs).
- **`$ai_span_id`** — Default: deterministic UUID from MCPCat event ID. Override to link MCPCat spans into an existing span hierarchy.
- **`$ai_span_name`** — Default: the MCP tool's `resourceName`. Override to add context like environment or region to the span label.

Tags that don't match a reserved `$ai_*` key are passed through as additional properties on the `$ai_span` event, queryable in PostHog like any other event property.

**Comment by naji247** [comment]
## Addressing review feedback from @kashishhora

**Datadog tag collision** (comment 1) — Good call. Addressed: customer tags are now prefixed with `mcpcat.`, keys sanitized (lowercase, strip colons/spaces/commas), values sanitized (strip commas).

**Sentry tag collision** (comment 2) — Also valid. Now addressed: customer tags in the Sentry exporter use the `mcpcat.` prefix. Sentry has ~55 reserved field names (`level`, `environment`, `release`, `server_name`, `transaction`, etc.) where custom tags would be shadowed by system fields in search. Sentry's docs explicitly recommend prefixing custom tags, and they render dot-separated names as a tree in the UI so `mcpcat.env` groups nicely.

**Datadog comma escaping** (comment 3) — Addressed: commas in tag values are replaced with `_` to prevent Datadog's ddtags parser from splitting them.

**Redaction protection** (comment 4) — Thanks for adding `tags` and `properties` to `PROTECTED_FIELDS` with the test coverage.

**Review by kashishhora — APPROVED** [review]
Lgtm! Let's get Rafa's thoughts on this :)

File: src/modules/exporters/posthog.ts [hunk 5]
```diff
@@ -114,6 +155,18 @@ export class PostHogExporter implements Exporter {
...
+    if (event.tags) {
+      for (const [key, value] of Object.entries(event.tags)) {
+        properties[`mcpcat_tag_${key}`] = value;
+      }
+    }
+
+    // Add customer-defined properties under a namespace
+    if (event.properties) {
+      properties.mcpcat_properties = event.properties;
+    }
```
Comment (rafaeelaudibert) [inline src/modules/exporters/posthog.ts]: This doesn't make sense for us. Nested properties are hard to use in PostHog - although supported. We'd prefer these were simply spreaded inside `properties`

```typescript
for (const [key, value] of Object.entries(event.properties)) {
  if (properties[key] !== undefined) properties[key] = value
} 
```

File: src/modules/exporters/posthog.ts [hunk 6]
```diff
@@ -114,6 +155,18 @@ export class PostHogExporter implements Exporter {
       properties.$set = $set;
     }
 
+    // Add customer-defined tags with prefix to avoid collisions
+    if (event.tags) {
+      for (const [key, value] of Object.entries(event.tags)) {
+        properties[`mcpcat_tag_${key}`] = value;
+      }
+    }
```
Comment (rafaeelaudibert) [inline src/modules/exporters/posthog.ts]: What's the rationale behind differentiating tags from properties?

We don't need to differentiate them, but I understand if they mean something inside MCPCat. In that case, this feels great!

File: src/modules/exporters/posthog.ts [hunk 7]
```diff
@@ -168,6 +218,58 @@ export class PostHogExporter implements Exporter {
...
+        properties[key] = value;
+      }
+    }
+
+    // Spread customer properties directly (can override MCPCat defaults)
+    if (event.properties) {
+      for (const [key, value] of Object.entries(event.properties)) {
+        properties[key] = value;
+      }
+    }
```
Comment (rafaeelaudibert) [inline src/modules/exporters/posthog.ts:264]: Hmmm, you're doing what I'm suggesting above in here - you're missing the check to avoid people overriding built-in keys, but that's fine. Also, tags do NOT include the `mcp_cat_` prefix. Any reason why these are different? They should all be the same IMO

**Review by rafaeelaudibert — CHANGES_REQUESTED** [review]
Reviewed the part that touches the posthog exporter, I don't have much context about the rest, but it seems to make sense!

Requesting changes because I'd prefer to avoid nested `mcp_cat_properties` JSON. They should just be spread in the base `properties`.

File: src/modules/exporters/posthog.ts (hunk 5 above)
Comment (naji247) [reply src/modules/exporters/posthog.ts]: No problem at all. Thanks for flagging 👍

File: src/modules/exporters/posthog.ts (hunk 6 above)
Comment (naji247) [reply src/modules/exporters/posthog.ts]: Yes! Good question! They are only useful for us and some of our broader roadmap. They're used the following way by MCPcat:

- tags: They have stricter type enforcement (key value pairs) and they are expected to be indexed and optimized for searching upon. If a user of MCPcat wants to filter by appended data quickly, they should use tags.

- properties: Less strict object format. Can be viewed on sessions for additional context but not actually searchable or indexed by MCPcat.

File: src/modules/exporters/posthog.ts (hunk 7 above)
Comment (naji247) [reply src/modules/exporters/posthog.ts:264]: I'll make them all the same. Just an oversight here :)
